From d2d720015813185d1ad4fa12884cab9bac6a8b25 Mon Sep 17 00:00:00 2001
From: Patrick Dawkins <pjcdawkins@gmail.com>
Date: Tue, 30 Aug 2016 11:53:52 +0100
Subject: [PATCH] Make it possible to get the access token without triggering a
 refresh

---
 src/Oauth2Subscriber.php | 6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

diff --git a/src/Oauth2Subscriber.php b/src/Oauth2Subscriber.php
index 67158bd..34e1509 100644
--- a/src/Oauth2Subscriber.php
+++ b/src/Oauth2Subscriber.php
@@ -110,16 +110,18 @@ public function onBefore(BeforeEvent $event)
     /**
      * Get the access token.
      *
+     * @param bool $refresh Whether to refresh the token, if possible.
+     *
      * @return AccessToken|null Oauth2 access token
      */
-    public function getAccessToken()
+    public function getAccessToken($refresh = true)
     {
         if ($this->accessToken && $this->accessToken->isExpired()) {
             // The access token has expired.
             $this->accessToken = null;
         }
 
-        if (null === $this->accessToken) {
+        if (null === $this->accessToken && $refresh) {
             // Try to acquire a new access token from the server.
             $this->accessToken = $this->acquireAccessToken();
             if ($this->accessToken) {
